repo/blob/s3/s3_storage.go

```diff
@@ -312,6 +312,45 @@ func (s *s3Storage) FlushCaches(ctx context.Context) error {
 	return nil
 }
 
+func isAccessDenied(err error) bool {
+	var e minio.ErrorResponse
+	return errors.As(err, &e) && e.Code == "AccessDenied"
+}
+
+func (s *s3Storage) bucketAccessible(ctx context.Context) error {
+	ok, err := s.cli.BucketExists(ctx, s.Options.BucketName)
+
+	if err == nil && ok {
+		return nil
+	} else if err == nil {
+		return errors.Errorf("bucket %q does not exist", s.Options.BucketName)
+	}
+
+	if !isAccessDenied(err) {
+		return errors.Wrap(err, "error checking bucket existence")
+	}
+
+	if s.Options.Prefix == "" {
+		return errors.Wrapf(err, "no permission to access bucket itself, bucket %q", s.Options.BucketName)
+	}
+
+	// Verify the existence of bucket by getting an non-existent object from the bucket/prefix and
+	// expecting an BlobNotFound error, in case that the account has no permission to the bucket itself
+	// but has permission to the prefix of the bucket
+	nonExistentBlob := fmt.Sprintf("kopia-s3-storage-initializing-%v", clock.Now().UnixNano())
+
+	var blobOutput gather.WriteBuffer
+	defer blobOutput.Close()
+
+	err = translateError(s.GetBlob(ctx, blob.ID(nonExistentBlob), 0, -1, &blobOutput))
+
+	if errors.Is(err, blob.ErrBlobNotFound) {
```

Comment:
I think the description from the document missed the factor of the prefix, in our scenarios it is like:

> If you have the s3:ListBucket permission on the bucket/prefix, Amazon S3 returns an HTTP status code 404 (Not Found) error.
> If you don’t have the s3:ListBucket permission on the bucket/prefix, Amazon S3 returns an HTTP status code 403 ("access denied") error.

At least, my local testing result is like this.
If so, this could meet our purpose, while 1 is our expectation the function returns `nil` for it:

1. If the account doesn't have permission to the bucket, but has permission to bucket/prefix, then 404 is returned
2. If the account doesn't have permission to the bucket and neither to bucket/prefix, then 403 is returned

Reply:
> Perhaps, if the error above (when checking for the bucket existence) is "access denied", simply log and return nil. Essentially, things "may or may not" work after that.

This is what the initial changes in this PR, while @jkowalski thought checking a non-existing object against the prefix  is better as it could also check whether the account has permission to the prefix dir. See here https://github.com/kopia/kopia/pull/3072#issuecomment-1596235663

Reply:
> This is what the initial changes in this PR, while @jkowalski thought checking a non-existing object against the prefix is better as it could also check the existence of the prefix dir. See here [#3072 (comment)](https://github.com/kopia/kopia/pull/3072#issuecomment-1596235663).

The suggestion is to check for a specific non-existing blob that should return "not found", because either: (1) the bucket does not exist; or (2) the blob does not exist. This does not check the existence of a prefix "directory" object.

At this point in this function, the return from `BucketExists(...)` is "access denied", which likely means the credential does not have permissions to list buckets or to list the name (not the contents) of this specific bucket.

According to the AWS S3 documentation, at this point doing a GET of a non-existent blob should always return "access denied" (here), and never "blob not found".

@Lyndon-Li Can you test this scenario?

- existing bucket
- credentials with:
  1. no access to listing this bucket, or all buckets
  2. access to listing, get, put etc of all blobs inside the bucket.
- perform a get of a non-existent blob with no prefix
- perform a get of a non-existent blob with an additional prefix

Also, what happens when the credential / principal does not have access to list the bucket and the bucket does not exist?
- Does `BucketExists(...)` return "access denied" or "not found"?
- What would the blob existence check return in that case?

Note:
There is no concept or distinction of directories in AWS S3, it is a flat key namespace. 
- An object named `dir/` is simply an object and it can have arbitrary contents.
- It is possible to have an object named `dir/object-1` without requiring the existence of another object with the same prefix, e.g. `dir`, or `dir/`.
This may not apply to other S3 "compatible" implementations.

Reply:
For sure, for an object store, there is no concept of dir. On the other hand, inside the AWS S3 policy, the permission of a bucket/prefix could be different from the bucket itself, like below configuration snippet:
        {
            "Effect": "Allow",
            "Action": [
                "s3:Get*",
                "s3:List*"
            ],
            "Resource": "arn:aws:s3:::bucket-123",
            "Condition": {
                "StringLike": {
                    "s3:prefix": "prefix-123/*"
                }
            }
        },
The above configuration means, the Get and List permission is only applied to `bucket-123/prefix-123`, which could be analogised as that the account has access to the sub-dir only.

Reply:
I have done the above tests through the current code in this PR, here are the results.
**no access to listing this bucket + perform a get of a non-existent blob with no prefix**
```
kopia repository create s3 --bucket=bucket-123
ERROR can't connect to storage: no permission to access bucket itself, bucket "bucket-123": Access Denied.
```

**no access to listing this bucket + perform a get of a non-existent blob with an additional prefix that the account has permission to access**
```
kopia repository create s3 --bucket=bucket-123 --prefix=prefix-123/test1/
Enter password to create new repository: 
Re-enter password for verification: 
Initializing repository with:
  block hash:          BLAKE2B-256-128
  encryption:          AES256-GCM-HMAC-SHA256
  splitter:            DYNAMIC-4M-BUZHASH
Connected to repository.
```

**no access to listing this bucket + perform a get of a non-existent blob with an additional prefix that the account has no permission to access**
```
kopia repository create s3 --bucket=bucket-123 --prefix=prefix-456/test1/
ERROR can't connect to storage: unable to determine if bucket dir is accessible, bucket "bucket-123", prefix "prefix-456/test1/": Access Denied
```

Reply:
Finally, I agree with this statement `This may not apply to other S3 "compatible" implementations`.
And though we have successfully tested the current code against AWS S3, we have no way to test it against all S3 compatible storages.

Therefore, looks like the preferred approach mentioned above is safer:

- Preferred: not perform the blob existence check at all when BucketExists(...) returns access denied. Simply log, and return nil. 

This will not able to capture **Test 3**, but it is totally OK, because later when the repository does further operations, the fail & abort will happen.

What do you think? @julio-lopez @jkowalski 
If you both agree, I will change the code accordingly.

Reply:
Hi @Lyndon-Li 
Thanks for running the tests, the explanation and your patience. This helped me understand the issues better.

After a conversation with Jarek, it seems that the best route is to simply remove the original bucket existence check and NOT perform any additional checks, that is, remove the helper altogether. The access check will be implicitly performed when getting the "storage config" blob below.

Thanks.

Reply:
Great! That is the simplest way, let me change the code accordingly.

Reply:
@julio-lopez @jkowalski 
I've changed the code accordingly. Please have another look.
Additionally, I didn't see any testing code is checking the return values of original bucket existence check, so I assume we don't need to change the testing code. And let me know any other thing I need to do in order to merge this PR.

Reply:
Here are the test results generated by the new code for the similar test cases above:
**Test 1: no access to this bucket + create repo to the bucket with no prefix**
```
kopia repository create s3 --bucket=bucket-123
ERROR can't connect to storage: error retrieving storage config from bucket "bucket-123": Access Denied
```

**Test 2: no access to this bucket + create repo with an additional prefix that the account has permission to access**
```
kopia repository create s3 --bucket=bucket-123 --prefix=prefix-123/test1/
Initializing repository with:
  block hash:          BLAKE2B-256-128
  encryption:          AES256-GCM-HMAC-SHA256
  splitter:            DYNAMIC-4M-BUZHASH
Connected to repository.
```

**Test 3: no access to listing this bucket + create a repo with an additional prefix that the account has no permission to access**
```
kopia repository create s3 --bucket=bucket-123 --prefix=prefix-456/test1/
ERROR can't connect to storage: error retrieving storage config from bucket "bucket-123": Access Denied
```